src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/SystemService.java

```diff
@@ -184,6 +190,82 @@ public AnchorFile getAnchorFile() throws AnchorNotFoundException {
         return anchorFile;
     }
 
+    /**
+     * Calculate the hex hash of the given anchor file. Used to verify/preview an anchor file before
+     * uploading it
+     * @param anchorBytes
+     * @return
+     * @throws InvalidAnchorInstanceException anchor is not generated in the current instance
+     */
+    public AnchorFile getAnchorFileFromBytes(byte[] anchorBytes) throws InvalidAnchorInstanceException {
+        ConfigurationAnchorV2 anchor = new ConfigurationAnchorV2(anchorBytes);
+        verifyAnchorInstance(anchor);
+        AnchorFile anchorFile = new AnchorFile(calculateAnchorHexHash(anchorBytes));
+        anchorFile.setCreatedAt(FormatUtils.fromDateToOffsetDateTime(anchor.getGeneratedAt()));
+        return anchorFile;
+    }
+
+    /**
+     * Upload a new configuration anchor. A temporary anchor file is created on the filesystem in order to run
+     * the verification process with configuration-client module (via external script).
+     * @param anchorBytes
+     * @throws InvalidAnchorInstanceException anchor is not generated in the current instance
+     * @throws AnchorUploadException in case of external process exceptions
+     */
+    public void uploadAnchor(byte[] anchorBytes) throws InvalidAnchorInstanceException, AnchorUploadException {
+        ConfigurationAnchorV2 anchor = new ConfigurationAnchorV2(anchorBytes);
+        verifyAnchorInstance(anchor);
+        File tempAnchor = null;
+        try {
+            tempAnchor = createTemporaryAnchorFile(anchorBytes);
+            configurationVerifier.verifyInternalConfiguration(tempAnchor.getAbsolutePath());
+            anchorRepository.saveAndReplace(tempAnchor);
+        } catch (ConfigurationVerifier.ConfigurationVerificationException ce) {
```

Comment:
I was thinking that the request itself was fine but there was something else wrong. I guess that this could actually be interpreted either way: as a bad request or internal error. Here are the translations for the different errors that `ConfigurationVerificationException` can mean:
```javascript
anchor_not_for_external_source: "Anchor points to an internal configuration source. Only external configuration source anchors are supported as trusted anchors"
missing_private_params: "Configuration anchor import failed: invalid anchor file"
unreachable: "Configuration source cannot be reached, check source URL in uploaded anchor file"
outdated: "Configuration from source is out of date"
signature_invalid: "Signature of configuration cannot be verified"
other: "Configuration from source failed verification"
```
For example `missing_private_params` could mean 400 but `other` could mean 500.

Reply:
Isn't missing_private_params like the others - anchor (such as internal-configuration-anchor.xml) itself was fine (even though old translation says it was invalid. Is the translation misleading?) but when app downloaded the internal configuration, it was corrupted (private params were missing)? So in that sense it would still be 500?

I kind of feel all of those should still be checked exceptions, just to communicate to the caller that this kind of things can go wrong, be prepared. Even if we will just http 500 them now, some future service could appreciate that when it builds on top of this method. But I am not really 100% on this, if you think runtime exceptions are better for these, go ahed.

Reply:
> but when app downloaded the internal configuration, it was corrupted (private params were missing)? So in that sense it would still be 500?

Don't know about this. I mean shouldn't the downloading itself be the one failing here already? 

> I kind of feel all of those should still be checked exceptions, just to communicate to the caller that this kind of things can go wrong, be prepared.

By this do you mean that all the different cases above (anchor_not_for_external_source, missing_private_params, unreachable etc.) should all actually be their own respected exceptions?

Reply:
> Don't know about this. I mean shouldn't the downloading itself be the one failing here already?

Not sure, one should read ConfigurationClientMain and verify_internal_configuration carefully to know for sure. For me it looks like missing_private_params could mean both, bad anchor or bad global conf, but I am not sure. If that is so, then maybe http 500 is more correct, if can't know if it is 400 or 500 for sure.

> By this do you mean that all the different cases above (anchor_not_for_external_source, missing_private_params, unreachable etc.) should all actually be their own respected exceptions?

No I think a shared checked exception class, like ConfigurationVerificationException but changed into checked one, would be enough.

Reply:
It is a checked exception. Just moved the handling to the controller layer and now `InternalServerErrorException` is explicitly thrown if `ConfigurationVerificationException` is caught. So basically it is now handled as a http 500 status since it is a bit unclear which one it should be